feat(block): support for having more than one title

samyosm opened the pull request:
To be able to create TUIs such as that of [btop](https://github.com/tui-rs-revival/ratatui/assets/99157490/2a7cbde2-f943-435b-96ea-07d4ddf94a67), I believe it to be necessary to be able to have multiple titles for a single block. This pull request adds the ability to have multiple titles per block. Each title can have a different alignment thought multiple titles can have the same alignment.

For backwards compatibility sake, I let the old title system and just added a new one that takes precedence over the old one in cases it has been set.

In the new system, you declare your titles in the following ways:
```rust
Block::default()
    .title("old system"),
    .titles(vec![
        Title::default()
            .content("Title 1")
            .alignment(Alignment::Right),
        Title::default()
            .content("Title 2")
            .alignment(Alignment::Right)
            .position(Position::Bottom),
        Title::default()
            .content("Title 3")
            .alignment(Alignment::Center)
            .position(Position::Bottom),
        Title::default()
            .content("Cen")
            .alignment(Alignment::Center)
            .position(Position::Bottom),
        Title::default()
            .content("Center 1")
            .alignment(Alignment::Center),
        Title::default().content("Cen").alignment(Alignment::Center),
    ])
    .borders(Borders::ALL);
```

where the content is just a `Line` like in the old system.

Note: I haven't added tests, documentation, or bothered with code style.

joshka commented:
Can you explain the need for this a bit more. How would you explain the feature to someone that hasn’t seen btop? I took a look at the screenshot and can’t identify how multiple titles is useful.

samyosm commented:
I believe it would allow for a better user experience. For example, similarly to Btop, the titles can be useful for indicating to users keyboard shortcuts they can make use of in a block; one might also want to add additional information in a block's border such as a status bar with word counts, current location, and such. Here are other uses I found:
1. Hierarchical information representation. E.g. a main title and a subtitle;
2. Navigation. E.g. arrows on each side of a block;
3. Step-by-step instructions or tutorials. I.e. something like a [wizard](https://uxdesign.cc/the-wizard-of-user-experience-6926ca41bc9a);
4. Tabs; and,
5. Filtering and sorting options.

joshka commented:
Sorry, I think I wasn't clear. Help me understand how the desired state is desirable. I'm looking for a description of the feature that tells me the effect of this change. How does it render? What are the features that it implements and how do they interact (multiple titles, alignment, on_bottom)? How does it interact with the current features of the block / borders?

For `on_bottom: bool`, idiomatic rust would suggest making this a position enum (Position::top/bottom) rather than a bool.

samyosm commented:
Oops, sorry for misunderstanding. I copied what was there before in the `Block` widget thinking that's how this library wanted it. Let me make the change real quick.

joshka commented:
Got it
Allow the block to render multiple titles which each can be positioned either top or bottom,  aligned (Left / Center / Right), and individually styled.

I wonder if we can change `title()` to accept `Into<Title>` instead of `Into<Line>` (and then `impl From<Line> for Title`). Then  we could allow title() to be called multiple times. e.g.:

```rust
Block::default()
    .title(Title::default()
        .content("Left")
        .alignment(Alignment::Left)
        .style(Style::default().fg(Color::Yellow))
    .title(Title::default().content("Center").alignment(Alignment::Center))
    .title(Title::default().content("Right").alignment(Alignment::Right))
    .title(Title::default().content("Left").alignment(Alignment::Left).position(Position::Bottom))
    .title(Title::default().content("Center").alignment(Alignment::Center)).position(Position::Bottom)
    .title(Title::default().content("Right").alignment(Alignment::Right)).position(Position::Bottom)
```
Thoughts?

samyosm commented:
Yes, true. I believe that to be a viable option. I'll work on it sometimes soon.

samyosm commented:
I implemented the suggestion you made but had to remove the old system's `title_alignment` and `title_on_bottom` functions. As such, I rewrote some demo examples and some tests.

I'm now uncertain whether or not to bring them back for backward compatibility. As reading them would mean people won't have to make any changes to their current library. We could just mark them deprecated. 

Please let me know your opinion on this.

joshka commented:
I think leave them in if possible and mark them deprecated.
Perhaps make the field used for the default for the alignment?

samyosm commented:
I added tests, and brought back `title_alignment` and `title_on_bottom`. All that is left is adding documentation and deprecating    `title_alignment` and `title_on_bottom`.

Backward compatibility-wise, people shouldn't have to change anything. Thought I haven't tested this.

joshka commented:
The examples are generally good for checking the backwards compatibility as they show what a user will see if they are using the deprecated methods. Assume the next major version for a deprecation (v0.22.0).

samyosm commented:
All done I believe.

joshka commented on src/widgets/block.rs:
```diff
@@ -311,8 +377,9 @@ impl<'a> Widget for Block<'a> {
                 .set_style(self.border_style);
         }
 
-        // Title
```
This method is getting long - can we extract the border rendering and the title rendering into two methods?

samyosm replied:
For now, I just separated them in two functions: `render_border` and `render_titles` in the block implementation. Later, I look for a better way of handling this.

joshka commented on tests/widgets_block.rs:
```diff
@@ -478,3 +480,151 @@ fn widgets_block_title_alignment_bottom() {
 }
+
+#[test]
+fn widgets_block_multiple_titles() {
```
We probably need more tests that show how things too long work and how multiple titles that would overlap are rendered.

samyosm replied:
For now overlapping titles render in the following way:

┌Title1─Title2
└───────┘
The titles don't wrap to a new line, but they cut the corner a bit. My guess is this is due to the ordering of the rendering. 

I would like to make it truncate or something else, but I'm now sure how to do that.

joshka reviewed (changes requested):
I'm generally approving the idea. Grouping the feedback from this review:
- consider whether to make the existing block methods act as defaults for each title. This makes it consistent with other widgets that do this (like `Paragraph::style()` which acts as a default)
- make sure the docs cover the edge cases (overlapping), note the defaults, and indicate how the titles will be rendered. (it might be useful to show a couple of example outputs in the docs to make it easy to visualize this)

samyosm commented on src/lib.rs:
```diff
@@ -181,6 +181,7 @@ pub mod style;
 pub mod symbols;
 pub mod terminal;
 pub mod text;
+pub mod title;
```
It isn't really a widget, more like an extension of the block widget.

joshka replied:
hrm, yeah I agree. Perhaps it belongs in block.rs for now?

samyosm replied:
So you believe we should put it in the block.rs file?

```rust
#[path = "../title.rs"]
pub mod title;
```

Or should we put title.rs inside the widget folder? (considering it isn't a widget).
Or maybe we should just let it like it is now.

joshka replied:
yeah, let's make this block specific to start with. I don't think we need a title anywhere else yet.

samyosm replied:
All done!

samyosm commented on examples/inline.rs:
```diff
@@ -226,9 +227,11 @@ fn run_app<B: Backend>(
-        .title("Progress")
-        .title_alignment(Alignment::Center);
+    let block = Block::default().title(
+        Title::default()
+            .content("Progress")
```
Thanks for reminding me that, I forgot to update after implementing `From`. I updated them.

samyosm commented on src/title.rs:
```diff
@@ -0,0 +1,91 @@
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Title<'a> {
+    pub content: Option<Line<'a>>,
```
Yeah, you're right it doesn't make sense. Just removed it. Though now the default is just an empty string.

samyosm commented on src/title.rs:
```diff
@@ -0,0 +1,91 @@
+    }
+}
+
+impl<'a> From<Line<'a>> for Title<'a> {
+    fn from(value: Line<'a>) -> Self {
+        Self::default().content(value)
+    }
+}
+
+impl<'a> From<String> for Title<'a> {
+    fn from(value: String) -> Self {
+        Self::default().content(value)
+    }
+}
+
+impl<'a> From<&'a str> for Title<'a> {
+    fn from(value: &'a str) -> Self {
+        Self::default().content(value)
+    }
+}
+
+impl<'a> From<Vec<Span<'a>>> for Title<'a> {
+    fn from(value: Vec<Span<'a>>) -> Self {
+        Self::default().content(value)
+    }
+}
+
+impl<'a> From<Span<'a>> for Title<'a> {
+    fn from(span: Span<'a>) -> Self {
+        Self::from(vec![span])
+    }
+}
```
Wow, that's a lot simpler way of doing it, thanks for showing me! I just made the change.

samyosm commented on src/widgets/block.rs:
```diff
@@ -96,15 +97,25 @@ impl Padding {
-    /// Whether or not title goes on top or bottom row of the block
-    title_on_bottom: bool,
+    /// Optional list of titles
+    titles: Option<Vec<Title<'a>>>,
```
Yes, it would make a lot more sense. I just made the change.

samyosm commented on src/widgets/block.rs:
```diff
@@ -121,9 +132,7 @@ pub struct Block<'a> {
 impl<'a> Default for Block<'a> {
     fn default() -> Block<'a> {
         Block {
-            title: None,
-            title_alignment: Alignment::Left,
-            title_on_bottom: false,
+            titles: None,
```
Yes, but wouldn't it make more sense to rename them `titles_alignment` and `titles_on_bottom`. Furthermore, if we are expecting users to have that many titles to need functions that act on all their titles, shouldn't we instead provide a `.titles` function that would accept a vector of titles. We could also accept position and alignment parameters in that function so that it would act on the vector the just gave.

## Example
### Approach 1
```rust
Block::default()
  .titles(vec!["A", "B"])
  .titles_with_alignement(vec!["C", "D"], Alignment::Left)
  .titles_with_position(vec!["E", "F"), Position::Bottom)
  .titles_with_all(vec!["G", "H"], Alignment::Right, Position::Bottom)
  .title(Title::from("I").position(Position::Top).alignment(Alignment::Left));
```

### Approach 2

```rust
Block::default()
  .titles(vec!["A", "B"])
  .titles(Titles::from(vec!["C", "D"]).alignment(Alignment::Left))
  .titles(Titles::from(vec!["E", "F"]).position(Position::Bottom))
  .titles(Titles::from(vec!["C", "D"])
        .alignment(Alignment::Left)
        .position(Position::Bottom))
  .title(Title::from("I").position(Position::Top).alignment(Alignment::Left));
```
Personally, I prefer approach 2.

joshka replied:
I think keeping something that works fine over making a small breaking change is probably the right choice here. I don't think the ergonomics of having both `title` and `titles` seems easy.

I'm looking at this from the lens of an existing user and also from the perspective of maintaining a library long-term. Rename -> breaking a user. Adding a new function -> too many items to choose from when using completion in your editor or reading docs.

I think making title_alignment() work on all titles is fine. Linguistically it makes sense as a singular even though it acts on multiple elements ("The title alignment of the block is Left" or "The block's title alignment is center"), and from there title_position flows naturally.

samyosm replied:
Sure, all right. Then I can make Title's `alignment` and `position` property optional, then make it so `title_alignment` and `title_position` changes all the alignments / positions of all titles except those that have already been set right?

samyosm replied:
I've just done it and added some documentation.

samyosm commented on src/widgets/block.rs:
```diff
@@ -134,33 +143,90 @@ impl<'a> Default for Block<'a> {
         self
     }
-
     #[deprecated(
         since = "0.10.0",
         note = "You should use styling capabilities of `text::Line` given as argument of the `title` method to apply styling to the title."
     )]
     pub fn title_style(mut self, style: Style) -> Block<'a> {
```
For the first code, we could have a `Titles` struct that would help developers deal with multiple title styling, alignment, and position. See [this comment](https://github.com/tui-rs-revival/ratatui/pull/232/files/e8c6464357cd0e41b668629028402ce633e2e237#r1225856916).

For the second code block, I'm not sure we should promote usage of deprecated functions. That might only lead us to work on old code instead of the new solutions we found. This is also the reason, I made `title_alignment` and `title_on_bottom` act on the first title only: to not promote their usage.

joshka replied:
Following from the previous. This would be removing a deprecation for a good reason. The behavior remains the same as it was originally (`Block::default().title("foo").title_style(s)` just works), while acting fine on multiple styles. I don't think the second block is good at all btw, but it highlights that we should note that this works as a default for all styles in the docs.

samyosm replied:
I'm not sure to have understood. Do you want the `title_style` method to act on all titles thus overriding previous `title_style` method invocation (in the second block, it would mean the only modifier is italic) or do you want `title_style` to work on the currently set titles (in the second block, it would mean title A is just Bold, and title B is just Italic).

samyosm replied:
To be more consistent with `title_position` and `title_alignment`, I made act on all the titles of the block. In the event a title already had some style, this method will [patch the style](https://docs.rs/ratatui/0.21.0/ratatui/text/struct.Line.html#method.patch_style) onto it.

samyosm commented on src/widgets/block.rs:
```diff
@@ -134,33 +143,90 @@ impl<'a> Default for Block<'a> {
         self
     }
 
+    #[deprecated(
+        since = "0.22.0",
+        note = "You should use a `Title` object to position a title."
+    )]
+    /// This method only positions the very first title
     pub fn title_on_bottom(mut self) -> Block<'a> {
```
Considering `title_on_bottom` is deprecated, I'm unsure we should create another deprecated function we do not even want people to use.

samyosm replied:
I've done it, in accordance with the previous talk.

samyosm commented on src/title.rs:
```diff
@@ -0,0 +1,91 @@
+    }
+}
+
+impl<'a> From<Title<'a>> for String {
```
https://github.com/tui-rs-revival/ratatui/blob/593fd29d00dde265be3dd83cc8c7ca132022cf87/src/text/line.rs#L129-L136
Not sure, I saw it implemented for Line so I thought it might be needed here as well. I removed it for now.

samyosm commented on src/widgets/block.rs:
```diff
@@ -134,33 +143,90 @@ impl<'a> Default for Block<'a> {
+    ///            .position(Position::Bottom),
+    ///    )
+    ///    .title(
+    ///        Title::default()
```
I cut down the number of titles in the example, and added one that doesn't provide a Title object. I also added the doc you made without the overlapping bit as I haven't test that yet.

samyosm commented on src/widgets/block.rs:
```diff
@@ -134,33 +143,90 @@ impl<'a> Default for Block<'a> {
         self
     }
 
+    #[deprecated(
+        since = "0.22.0",
+        note = "You should use a `Title` object to align the title(s)."
+    )]
+    /// This method only aligns the very first title
     pub fn title_alignment(mut self, alignment: Alignment) -> Block<'a> {
```
This issue to is related to [this comment](https://github.com/tui-rs-revival/ratatui/pull/232/files/e8c6464357cd0e41b668629028402ce633e2e237#r1225856916).

joshka commented:
A lot of my comments come down to picking an API that:
- avoids breaking changes when possible and reasonable
- is the simplest least surprising interface
- avoids repetition

I think that 
```rust
Block::new()
  .title("A")
  .title("B")
  .title_style(Style::default()...)
  .title_position(Position::Bottom)
```
is fine

samyosm commented on src/title.rs:
```diff
@@ -0,0 +1,91 @@
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Title<'a> {
```
All instances of `Title::default` have been replaced to `Title:from` both in code and documentation.

joshka reviewed (changes requested):
Still looking for some changes on this to make it easy to see how it works.

The main change needed is to render any title style, position or alignment as an override in the render rather than by modifying the stored value.

joshka commented:
P.s. Thanks for sticking with it on this review.

I think the next change you make will probably be ready to merge. When you push can you please squash the changes into a single commit and update the commit message to use the conventional commit format with a header and a body that describes the feature enough that it works well for the changelog.

I didn't mention in the review, but would like to see it some tests that show how overlapping titles are handled (e.g. what happens when there is only 10 characters wide and there are 3 titles (left, center, right aligned) and they are each 4 characters wide...?)

samyosm commented on src/widgets/block.rs:
```diff
@@ -310,35 +372,100 @@ impl<'a> Widget for Block<'a> {
                 .set_style(self.border_style);
         }
+    }
+    fn render_titles(&self, area: Rect, buf: &mut Buffer) {
```
Sure, I've just redone it with a similar approach to yours. I will now be adding some tests.

samyosm replied:
Actually, I decided to simply do:
```rust
fn render_title_position(&self, position: Position, area: Rect, buf: &mut Buffer) {
    // Note: the order in which these functions are called define the overlapping behavior
    self.render_right_titles(position, area, buf);
    self.render_center_titles(position, area, buf);
    self.render_left_titles(position, area, buf);
}

fn render_titles(&self, area: Rect, buf: &mut Buffer) {
    self.render_title_position(Position::Top, area, buf);
    self.render_title_position(Position::Bottom, area, buf);
}
```

for the following reasons:
- Easier to debug overlapping behavior
- Easier to change the math behind the calculation of an alignment
- Hashmaps became [weird](https://users.rust-lang.org/t/unexpected-random-behavior-in-hashmap/68416)

for the following inconveniences:
- Somewhat slower (unnoticeable as there won't be that many titles)

samyosm commented:
I tried using [Commitizen](https://github.com/lintingzhen/commitizen-go) for the commit, message but it would seem not to have worked.

joshka commented on tests/widgets_block.rs:
```diff
@@ -46,6 +49,80 @@ fn widgets_block_renders() {
+            width: 11,
+            height: 1,
+        },
+        Buffer::with_lines(vec!["aaaaabaaaaa"]),
```
This is a really odd result!
It happens because we're not clearing the space between each title.
I'm not sure it's worth fixing. We could probably leave this and worry about it if someone actually ever hits it.

samyosm replied:
What do you believe should happen instead?

joshka replied:
I'm not 100% certain, but probably here "aaaaa aaaaa" would be kinda right?

samyosm replied:
Well the behavior I set was the following: left titles override the center titles which override the right alignments. As such, when the space is too small, people might be able to grasp where other titles are if they aren't overwritten. I didn't want to declare spaces between titles as it might remove some information for the user.

joshka replied:
Yeah, I think this is the sort of thing that we can probably defer until someone gets impacted by it and work it out based on a real world application.

joshka replied:
yep - it happens to the best of us :D

joshka reviewed (changes requested):
Please rebase on main (there's an updated typos.toml that has a fix for the ratatui typo) and fix up the typos (use [`typos --write-changes`](https://github.com/crate-ci/typos/)?

samyosm commented on src/widgets/block.rs:
```diff
@@ -310,35 +374,120 @@ impl<'a> Widget for Block<'a> {
+            && title.position.unwrap_or(self.titles_position) == position
+    }
+
+    fn get_dx(&self, area: Rect) -> (u16, u16, u16) {
```
Yeah, you are right, I renamed it to `calculate_title_area_offsets`.

samyosm commented:
Wait, did I mess something up while rebasing? I'm new to this and I'm not sure.

joshka commented:
Yes maybe, but it's not the end of the world.
If you want to get back to exactly where you were, take a look at the output of `git reflog` There will a line that is just before a line that says `rebase (start): checkout main`. You can write `git reset --hard <the SHA1 hash on that line>` and you're exactly where you were.

I'd recommend taking the following steps:
```
git switch main

# assuming the this repo is your upstream remote:
git pull upstream main 

git switch multiple-titles

# this is optional - it just creates a branch that helps you easily go back to where you were
git branch mutliple-titles-before-rebase 

# assuming that you've already reset the branch back to where you want to merge
git rebase -i <some commit hash that is not your code>
# edit each line that is your code except the first and mark it as s or squash. close the file
# this will result in the code that you've written being a single commit
# now rebase on main

git rebase main
git push origin multiple-titles --force
```

samyosm commented:
Thanks a lot, that really helped.
